### PR TITLE
Support SD3 controlnet inpainting

### PR DESCRIPTION
**Pull request by JPlin**
# What does this PR do?

This PR support controlnet inpaint pipeline in stable diffusion 3, which contains specific weight for controlnet inpainting.

We hope this PR can be an initial version for StableDiffusion3ControlNetInpaintingPipeline

We put the demo here: [DEMO](https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting).

## How to use it?
Examples:
```py
import torch
from diffusers.utils import load_image, check_min_version
from diffusers.pipelines import StableDiffusion3ControlNetInpaintingPipeline
from diffusers.models.controlnet_sd3 import SD3ControlNetModel

check_min_version("0.30.0.dev0")

controlnet = SD3ControlNetModel.from_pretrained(
    "alimama-creative/SD3-Controlnet-Inpainting",
    use_safetensors=True,
    extra_conditioning_channels=1
)
pipe = StableDiffusion3ControlNetInpaintingPipeline.from_pretrained(
    "stabilityai/stable-diffusion-3-medium-diffusers",
    controlnet=controlnet,
    torch_dtype=torch.float16,
)
pipe.text_encoder.to(torch.float16)
pipe.controlnet.to(torch.float16)
pipe.to("cuda")

image = load_image("https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting/blob/main/images/prod.png")
mask = load_image("https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting/blob/main/images/mask.jpeg")
width = 1024
height = 1024
prompt="a woman wearing a white jacket, black hat and black pants is standing in a field, the hat writes SD3"
generator = torch.Generator(device="cuda").manual_seed(24)
res_image = pipe(
    negative_prompt='deformed, distorted, disfigured, poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, mutated hands and fingers, disconnected limbs, mutation, mutated, ugly, disgusting, blurry, amputation, NSFW',
    prompt=prompt,
    height=height,
    width=width,
    control_image = image,
    control_mask = mask,
    num_inference_steps=28,
    generator=generator,
    controlnet_conditioning_scale=0.95,
    guidance_scale=7,
).images[0]
res_image.save(f'sd3.png')
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@JPlin thank you!

@asomoza can you check this out?

**Comment by asomoza** [comment]
Thanks a lot for your work!

IMO the example images doesn't do justice on how good this is. I suggest you use some better quality images and with 1024px resolution.

This one probably is going to become my personal favorite for inpainting.

As an example I'm going to use my dog example which is really hard for inpainting:

|source|mask|result|
|---|---|---|
|ComfyUI_00225_|dog_mask||

The best part is that this is not cherry picked, some more examples:

| | | |
|---|---|---|
||||

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 1]
```diff
@@ -0,0 +1,1137 @@
...
+                image=control_image,
+                mask=control_mask,
+                width=width,
+                height=height,
+                batch_size=batch_size * num_images_per_prompt,
+                num_images_per_prompt=num_images_per_prompt,
+                device=device,
+                dtype=dtype,
+                do_classifier_free_guidance=self.do_classifier_free_guidance,
+                guess_mode=False,
```
Comment (asomoza) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:1002]: why didn't you implement guess mode? I think this would be really good for this controlnet.

**Review by asomoza — APPROVED** [review]
LGTM, just left one comment

**Comment by asomoza** [comment]
you need to run `make style` and `make quality` to fix the test errors.

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py (hunk 1 above)
Comment (JPlin) [reply src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:1002]: We achieved results with guess mode = True; however, it is a conjecture that the lack of an encoding phase in the structure of the Dit block sequence, unlike UNet, may be responsible for some artifacts and checkerboard effects in the results. Therefore, keeping guess mode = False is more appropriate for now.

**Comment by asomoza** [comment]
can you add the pipeline to the [docs](https://huggingface.co/docs/diffusers/api/pipelines/controlnet_sd3) please.

**Comment by JPlin** [comment]
> can you add the pipeline to the [docs](https://huggingface.co/docs/diffusers/api/pipelines/controlnet_sd3) please.

Done！I update the docs/source/en/api/pipelines/controlnet_sd3.md file.

**Comment by asomoza** [comment]
can you fix the failing tests again please and merge with the main branch please.

**Comment by asomoza** [comment]
@a-r-r-o-w can you give this PR a review please.

File: docs/source/en/api/pipelines/controlnet_sd3.md [hunk 2]
```diff
@@ -22,7 +22,14 @@ The abstract from the paper is:
...
 
-This code is implemented by [The InstantX Team](https://huggingface.co/InstantX). You can find pre-trained checkpoints for SD3-ControlNet on [The InstantX Team](https://huggingface.co/InstantX) Hub profile.
+This controlnet code is mainly implemented by [The InstantX Team](https://huggingface.co/InstantX). The inpainting-related code was developed by [The Alimama Creative Team](https://huggingface.co/alimama-creative). You can find pre-trained checkpoints for SD3-ControlNet in the table below: 
+
+
+| control type | developer | link |
+| --------| ----------| ---- |
+| canny & pose & tile | [The InstantX Team](https://huggingface.co/InstantX) | [link](https://huggingface.co/InstantX)
+inpainting | [The AlimamaCreative Team](https://huggingface.co/alimama-creative) | [link](https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting) |
```
Comment (a-r-r-o-w) [inline docs/source/en/api/pipelines/controlnet_sd3.md:31]:
```suggestion
| ControlNet type | Developer | Link |
| -------- | ---------- | ---- |
| Canny | [The InstantX Team](https://huggingface.co/InstantX) | [Link](https://huggingface.co/InstantX/SD3-Controlnet-Canny) |
| Pose | [The InstantX Team](https://huggingface.co/InstantX) | [Link](https://huggingface.co/InstantX/SD3-Controlnet-Pose) |
| Tile | [The InstantX Team](https://huggingface.co/InstantX) | [Link](https://huggingface.co/InstantX/SD3-Controlnet-Tile) |
| Inpainting | [The AlimamaCreative Team](https://huggingface.co/alimama-creative) | [link](https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting) |
```

Would prefer full links and individual mentions here

File: docs/source/en/api/pipelines/controlnet_sd3.md [hunk 3]
```diff
@@ -35,5 +42,10 @@ Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers)
 	- all
 	- __call__
 
+## StableDiffusion3ControlNetInpaintingPipeline
+[[autodoc]] StableDiffusion3ControlNetInpaintingPipeline
```
Comment (a-r-r-o-w) [inline docs/source/en/api/pipelines/controlnet_sd3.md:46]:
```suggestion
[[autodoc]] pipelines.controlnet_sd3.pipeline_stable_diffusion_3_controlnet_inpainting.StableDiffusion3ControlNetInpaintingPipeline
```
To fix the docs related tests

File: src/diffusers/models/controlnet_sd3.py [hunk 4]
```diff
@@ -55,6 +55,7 @@ def __init__(
         pooled_projection_dim: int = 2048,
         out_channels: int = 16,
         pos_embed_max_size: int = 96,
+        extra_conditioning_channels: int = 0,
```
Comment (a-r-r-o-w) [inline src/diffusers/models/controlnet_sd3.py:58]: Instead of adding an extra parameter here, could we update the inpainting controlnet model config to have the correct value of `in_channels` directly? I don't think adding an extra parameter would be okay but will leave the decision to @sayakpaul or @yiyixuxu

File: src/diffusers/pipelines/__init__.py [hunk 5]
```diff
@@ -170,6 +170,7 @@
     _import_structure["controlnet_sd3"].extend(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/__init__.py:170]: The `__init__.py` files looks correct. I think you will have to update one more file: `src/diffusers/__init__.py`

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 6]
```diff
@@ -0,0 +1,1139 @@
...
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers.utils import load_image, check_min_version
+        >>> from diffusers.pipelines import StableDiffusion3ControlNetInpaintingPipeline
+        >>> from diffusers.models.controlnet_sd3 import SD3ControlNetModel
+
+        >>> check_min_version("0.30.0.dev0")
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:63]:
```suggestion
```

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 7]
```diff
@@ -0,0 +1,1139 @@
...
+
+        >>> controlnet = SD3ControlNetModel.from_pretrained(
+        >>>     "alimama-creative/SD3-Controlnet-Inpainting",
+        >>>     use_safetensors=True,
+        >>>     extra_conditioning_channels=1
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:68]: Let's set the `in_channels` to 17 instead of 16 here: https://huggingface.co/alimama-creative/SD3-Controlnet-Inpainting/blob/main/config.json?

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 8]
```diff
@@ -0,0 +1,1139 @@
...
+                unscale_lora_layers(self.text_encoder, lora_scale)
+
+        if self.text_encoder_2 is not None:
+            if isinstance(self, SD3LoraLoaderMixin) and USE_PEFT_BACKEND:
+                # Retrieve the original scale by scaling back the LoRA layers
+                unscale_lora_layers(self.text_encoder_2, lora_scale)
+
+        return prompt_embeds, negative_prompt_embeds, pooled_prompt_embeds, negative_pooled_prompt_embeds
+
+    def check_inputs(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:562]: I think this is similar to other SD3 pipelines, no? If so, could you add a `# Copied from` statement here?

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 9]
```diff
@@ -0,0 +1,1139 @@
...
+        prompt_3: Optional[Union[str, List[str]]] = None,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        num_inference_steps: int = 28,
+        timesteps: List[int] = None,
+        guidance_scale: float = 7.0,
+        control_guidance_start: Union[float, List[float]] = 0.0,
+        control_guidance_end: Union[float, List[float]] = 1.0,
+        control_image: PipelineImageInput = None,
+        control_mask: PipelineImageInput = None,
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:785]: The documentation for the new parameters are missing

**Review by a-r-r-o-w — COMMENTED** [review]
Thank you for the amazing work here with the new controlnet! The PR is great and implementation is really nice overall - just have a few minor nits. The results are very cool too! Let's get this merged soon

Really like the name "Alimama" :tada:

File: src/diffusers/models/controlnet_sd3.py (hunk 4 above)
Comment (JPlin) [reply src/diffusers/models/controlnet_sd3.py:58]: There are two modules,  `pos_embed` and `pos_embed_input`, that utilize this parameter. For `pos_embed`, the value is fixed, while for `pos_embed_input`, adding this parameter allows for changing the number of channels based on different control types. Therefore, I think it is reasonable to add an additional parameter.

**Comment by JPlin** [comment]
> can you fix the failing tests again please and merge with the main branch please.

image
I tried to run make quality and make style locally, but a lot of formatting issues were detected in files that I haven't changed. How should I handle these?

**Comment by a-r-r-o-w** [comment]
> I tried to run make quality and make style locally, but a lot of formatting issues were detected in files that I haven't changed. How should I handle these?

This mostly happens due to differing ruff versions I believe. Could you install `ruff==0.1.5` and try? If it still doesn't work, I can push the relevant changes

**Comment by JPlin** [comment]
> > I tried to run make quality and make style locally, but a lot of formatting issues were detected in files that I haven't changed. How should I handle these?
> 
> This mostly happens due to differing ruff versions I believe. Could you install `ruff==0.1.5` and try? If it still doesn't work, I can push the relevant changes

It works. Thank you for your code comments and suggestions. I have [make style & make quality](https://github.com/huggingface/diffusers/pull/9099/commits/92a57f6da51d4f1feb47a2d4d48638558c425168)

File: src/diffusers/models/controlnet_sd3.py (hunk 4 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/controlnet_sd3.py:58]: @sayakpaul @DN6 WDYT about this change? I think it should be okay, just wanted to confirm. The PR looks good to merge otherwise

Comment (sayakpaul) [reply src/diffusers/models/controlnet_sd3.py:58]: Should be fine IMO.

Comment (JPlin) [reply src/diffusers/models/controlnet_sd3.py:58]: Thx, o I understand that the current changes are okay. Can I resolve this conversation?

**Comment by JPlin** [comment]
Could you please review this request? @yiyixuxu

File: src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py [hunk 10]
```diff
@@ -0,0 +1,1144 @@
...
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers.utils import load_image, check_min_version
+        >>> from diffusers.pipelines import StableDiffusion3ControlNetInpaintingPipeline
+        >>> from diffusers.models.controlnet_sd3 import SD3ControlNetModel
+
+        >>> controlnet = SD3ControlNetModel.from_pretrained(
+        ...     "alimama-creative/SD3-Controlnet-Inpainting", use_safetensors=True, extra_conditioning_channels=1
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:64]:
```suggestion
        ...     "alimama-creative/SD3-Controlnet-Inpainting", use_safetensors=True
```
Let's add `extra_conditioning_channels: 1` in the config.json file at you checkpoint.

Comment (JPlin) [reply src/diffusers/pipelines/controlnet_sd3/pipeline_stable_diffusion_3_controlnet_inpainting.py:64]: Add extra_conditioning_channels: 1 in the config.json file at HF checkpoint.

**Comment by a-r-r-o-w** [comment]
@JPlin Could you add a inpainting test `test_controlnet_sd3_inpainting.py` similar to https://github.com/huggingface/diffusers/blob/main/tests/pipelines/controlnet_sd3/test_controlnet_sd3.py in the same folder?

You don't need to add slow tests, just the fast tests will suffice. You will have to update the get_dummy_components method with appropriate parameters to reflect a tiny inpainting controlnet, and make use of mask input in your tests. Let me know if you need any help with this. The PR looks ready to merge otherwise!

**Comment by JPlin** [comment]
> @JPlin Could you add a inpainting test `test_controlnet_sd3_inpainting.py` similar to https://github.com/huggingface/diffusers/blob/main/tests/pipelines/controlnet_sd3/test_controlnet_sd3.py in the same folder?
> 
> You don't need to add slow tests, just the fast tests will suffice. You will have to update the get_dummy_components method with appropriate parameters to reflect a tiny inpainting controlnet, and make use of mask input in your tests. Let me know if you need any help with this. The PR looks ready to merge otherwise!

Done!
